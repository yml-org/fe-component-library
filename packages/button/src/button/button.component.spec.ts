import { LitElement } from 'lit-element';

describe('button-component', () => {
  const BUTTON_COMPONENT = 'button-component';
  let buttonElement: LitElement;

  const getShadowRoot = (tagName: string): ShadowRoot =>
    document.body.getElementsByTagName(tagName)[0].shadowRoot;

  beforeEach(() => {
    buttonElement = window.document.createElement(
      BUTTON_COMPONENT
    ) as LitElement;
    document.body.appendChild(buttonElement);
  });

  afterEach(() => {
    document.body.getElementsByTagName(BUTTON_COMPONENT)[0].remove();
  });

  it('matches component snapshot', async () => {
    expect(
      document.body.getElementsByTagName(BUTTON_COMPONENT)[0]
    ).toMatchSnapshot();
  });

  it('sets button type', async () => {
    window.document.body.appendChild(buttonElement);
    await buttonElement.updateComplete;
    const buttonElementType = getShadowRoot(BUTTON_COMPONENT).querySelector('button').type;
      expect(buttonElementType).toBe('button');
  });

  it('sets button text', async () => {
    buttonElement['text']="YML";
    window.document.body.appendChild(buttonElement);
    await buttonElement.updateComplete;
    const buttonElementText = getShadowRoot(BUTTON_COMPONENT).querySelector('button').textContent;
      expect(buttonElementText).toContain( "YML");
  });

  it('sets button variant as contained', async () => {
    const buttonEle = getShadowRoot(BUTTON_COMPONENT).querySelector('button'); 
    expect(buttonEle.classList.contains('focus:bg-blue-700')).toBe(true);
  });

  it('sets button color as primary', async () => {
    const buttonEle = getShadowRoot(BUTTON_COMPONENT).querySelector('button'); 
     expect(buttonEle.classList.contains('bg-blue-600')).toBe(true);
  });
  

  it('sets button color as secondary', async () => {
    buttonElement['color']="secondary";
    window.document.body.appendChild(buttonElement);
    await buttonElement.updateComplete;
    const buttonEle = getShadowRoot(BUTTON_COMPONENT).querySelector('button'); 
    expect(buttonEle.classList.contains('bg-purple-600')).toBe(true);
  });

  it('sets button color as success', async () => {
    buttonElement['color']="success";
    window.document.body.appendChild(buttonElement);
    await buttonElement.updateComplete;
    const buttonEle = getShadowRoot(BUTTON_COMPONENT).querySelector('button'); 
    expect(buttonEle.classList.contains('bg-green-500')).toBe(true);
  });

  it('sets button color as danger', async () => {
    buttonElement['color']="danger";
    window.document.body.appendChild(buttonElement);
    await buttonElement.updateComplete;
    const buttonEle = getShadowRoot(BUTTON_COMPONENT).querySelector('button'); 
    expect(buttonEle.classList.contains('bg-red-600')).toBe(true);
  });

  it('sets button color as warning', async () => {
    buttonElement['color']="warning";
    window.document.body.appendChild(buttonElement);
    await buttonElement.updateComplete;
    const buttonEle = getShadowRoot(BUTTON_COMPONENT).querySelector('button'); 
    expect(buttonEle.classList.contains('bg-yellow-500')).toBe(true);
  });

  it('sets button color as light', async () => {
    buttonElement['color']="light";
    window.document.body.appendChild(buttonElement);
    await buttonElement.updateComplete;
    const buttonEle = getShadowRoot(BUTTON_COMPONENT).querySelector('button'); 
    expect(buttonEle.classList.contains('bg-gray-200')).toBe(true);
  });

  it('sets button color as dark', async () => {
    buttonElement['color']="dark";
    window.document.body.appendChild(buttonElement);
    await buttonElement.updateComplete;
    const buttonEle = getShadowRoot(BUTTON_COMPONENT).querySelector('button'); 
    expect(buttonEle.classList.contains('bg-gray-800')).toBe(true);
  });

  it('sets button variant as outlined', async () => {
    buttonElement['variant']="outlined";
    window.document.body.appendChild(buttonElement);
    await buttonElement.updateComplete;
    const buttonEle = getShadowRoot(BUTTON_COMPONENT).querySelector('button'); 
    expect(buttonEle.classList.contains('border-2')).toBe(true);
  });

  it('sets button variant as link', async () => {
    buttonElement['variant']="link";
    window.document.body.appendChild(buttonElement);
    await buttonElement.updateComplete;
    const buttonEle = getShadowRoot(BUTTON_COMPONENT).querySelector('button'); 
    expect(buttonEle.classList.contains('bg-transparent')).toBe(true);
  });

  it('sets the isDisabled property correctly', async () => {
    buttonElement['isDisabled']= true;
    window.document.body.appendChild(buttonElement);
    await buttonElement.updateComplete;
    const buttonEle = getShadowRoot(BUTTON_COMPONENT).querySelector('button'); 
    expect(buttonEle.classList.contains('pointer-events-none')).toBe(true);
  });

  it('sets the isDisabled property to false by default', async () => {
    const buttonEle = getShadowRoot(BUTTON_COMPONENT).querySelector('button'); 
    expect(buttonEle.classList.contains('pointer-events-none')).toBe(false);
  });

  it('sets the fullWidth property correctly', async () => {
    buttonElement['fullWidth']= true;
    window.document.body.appendChild(buttonElement);
    await buttonElement.updateComplete;
    const buttonEle = getShadowRoot(BUTTON_COMPONENT).querySelector('button'); 
    expect(buttonEle.classList.contains('w-full')).toBe(true);
  });

  it('sets the fullWidth property to false, by default', async () => {
    const buttonEle = getShadowRoot(BUTTON_COMPONENT).querySelector('button'); 
    expect(buttonEle.classList.contains('w-full')).toBe(false);
  });
  
  it('sets the isRounded property correctly', async () => {
    buttonElement['isRounded']= true;
    window.document.body.appendChild(buttonElement);
    await buttonElement.updateComplete;
    const buttonEle = getShadowRoot(BUTTON_COMPONENT).querySelector('button'); 
    expect(buttonEle.classList.contains('rounded-full')).toBe(true);
  });

  it('renders the component passed in slot', async () => {
    buttonElement['slotName'] = "btn";
    window.document.body.appendChild(buttonElement);
    document.querySelector('button-component').innerHTML =
      '<h2 slot="btn" class="slot-test">Test</h2>';
    await buttonElement.updateComplete;
    const elem = document
      .querySelector('button-component')
      .querySelector('h2');
    expect(elem.classList.contains('slot-test')).toBe(true);
  });
  
  it('sets part property correctly', async () => {
    const elem = getShadowRoot(BUTTON_COMPONENT).querySelector('button');
    expect(elem.part[0]).toBe("custom-button");
  });

  it('sets role property correctly', async () => {
    const elem = getShadowRoot(BUTTON_COMPONENT).querySelector('button');
    expect(elem.getAttribute('role')).toBe("button");
  });

});

