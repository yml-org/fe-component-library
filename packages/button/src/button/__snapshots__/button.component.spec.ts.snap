// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button-component matches component snapshot 1`] = `
u {
  "_$AL": Map {},
  "_$Do": O {
    "_$AA": <!---->,
    "_$AB": null,
    "_$AH": yt {
      "_$AD": N {
        "el": <template />,
        "parts": Array [
          Object {
            "ctor": [Function],
            "index": 0,
            "name": "tabindex",
            "strings": Array [
              "",
              "",
            ],
            "type": 1,
          },
          Object {
            "ctor": [Function],
            "index": 0,
            "name": "role",
            "strings": Array [
              "",
              "",
            ],
            "type": 1,
          },
          Object {
            "ctor": [Function],
            "index": 0,
            "name": "class",
            "strings": Array [
              "",
              "",
            ],
            "type": 1,
          },
          Object {
            "index": 1,
            "type": 2,
          },
          Object {
            "ctor": [Function],
            "index": 2,
            "name": "name",
            "strings": Array [
              "",
              "",
            ],
            "type": 1,
          },
        ],
      },
      "_$AM": [Circular],
      "_$AN": undefined,
      "u": Array [
        R {
          "_$AH": "0",
          "_$AM": [Circular],
          "_$AN": undefined,
          "element": <button
            class="inline-block px-6 font-medium text-xs leading-tight uppercase focus:outline-none focus:ring-0 transition duration-150 ease-in-out
      py-2.5 bg-blue-600 text-white shadow-md hover:bg-blue-700 hover:shadow-lg focus:bg-blue-700 focus:shadow-lg  active:bg-blue-800 active:shadow-lg
      rounded
      undefined
      undefined
    "
            part="custom-button"
            role="button"
            tabindex="0"
            type="button"
          >
            
        
            <!--?lit$7157807873$-->
            
        
            <slot
              name=""
            />
            
      
          </button>,
          "name": "tabindex",
          "options": Object {
            "host": [Circular],
            "isConnected": true,
            "renderBefore": null,
          },
          "type": 1,
        },
        R {
          "_$AH": "button",
          "_$AM": [Circular],
          "_$AN": undefined,
          "element": <button
            class="inline-block px-6 font-medium text-xs leading-tight uppercase focus:outline-none focus:ring-0 transition duration-150 ease-in-out
      py-2.5 bg-blue-600 text-white shadow-md hover:bg-blue-700 hover:shadow-lg focus:bg-blue-700 focus:shadow-lg  active:bg-blue-800 active:shadow-lg
      rounded
      undefined
      undefined
    "
            part="custom-button"
            role="button"
            tabindex="0"
            type="button"
          >
            
        
            <!--?lit$7157807873$-->
            
        
            <slot
              name=""
            />
            
      
          </button>,
          "name": "role",
          "options": Object {
            "host": [Circular],
            "isConnected": true,
            "renderBefore": null,
          },
          "type": 1,
        },
        R {
          "_$AH": "inline-block px-6 font-medium text-xs leading-tight uppercase focus:outline-none focus:ring-0 transition duration-150 ease-in-out
      py-2.5 bg-blue-600 text-white shadow-md hover:bg-blue-700 hover:shadow-lg focus:bg-blue-700 focus:shadow-lg  active:bg-blue-800 active:shadow-lg
      rounded
      undefined
      undefined
    ",
          "_$AM": [Circular],
          "_$AN": undefined,
          "element": <button
            class="inline-block px-6 font-medium text-xs leading-tight uppercase focus:outline-none focus:ring-0 transition duration-150 ease-in-out
      py-2.5 bg-blue-600 text-white shadow-md hover:bg-blue-700 hover:shadow-lg focus:bg-blue-700 focus:shadow-lg  active:bg-blue-800 active:shadow-lg
      rounded
      undefined
      undefined
    "
            part="custom-button"
            role="button"
            tabindex="0"
            type="button"
          >
            
        
            <!--?lit$7157807873$-->
            
        
            <slot
              name=""
            />
            
      
          </button>,
          "name": "class",
          "options": Object {
            "host": [Circular],
            "isConnected": true,
            "renderBefore": null,
          },
          "type": 1,
        },
        O {
          "_$AA": <!--?lit$7157807873$-->,
          "_$AB": 
        ,
          "_$AH": Symbol(lit-nothing),
          "_$AM": [Circular],
          "_$AN": undefined,
          "_$Cm": true,
          "options": Object {
            "host": [Circular],
            "isConnected": true,
            "renderBefore": null,
          },
          "type": 2,
        },
        R {
          "_$AH": undefined,
          "_$AM": [Circular],
          "_$AN": undefined,
          "element": <slot
            name=""
          />,
          "name": "name",
          "options": Object {
            "host": [Circular],
            "isConnected": true,
            "renderBefore": null,
          },
          "type": 1,
        },
      ],
    },
    "_$AM": undefined,
    "_$AN": undefined,
    "_$Cm": true,
    "options": Object {
      "host": [Circular],
      "isConnected": true,
      "renderBefore": null,
    },
    "type": 2,
  },
  "_$E_": Promise {},
  "_$Ei": undefined,
  "_$El": null,
  "enableUpdating": [Function],
  "hasUpdated": true,
  "isUpdatePending": false,
  "renderOptions": Object {
    "host": [Circular],
    "isConnected": true,
    "renderBefore": null,
  },
  "renderRoot": ShadowRoot {
    "_$litPart$": O {
      "_$AA": <!---->,
      "_$AB": null,
      "_$AH": yt {
        "_$AD": N {
          "el": <template />,
          "parts": Array [
            Object {
              "ctor": [Function],
              "index": 0,
              "name": "tabindex",
              "strings": Array [
                "",
                "",
              ],
              "type": 1,
            },
            Object {
              "ctor": [Function],
              "index": 0,
              "name": "role",
              "strings": Array [
                "",
                "",
              ],
              "type": 1,
            },
            Object {
              "ctor": [Function],
              "index": 0,
              "name": "class",
              "strings": Array [
                "",
                "",
              ],
              "type": 1,
            },
            Object {
              "index": 1,
              "type": 2,
            },
            Object {
              "ctor": [Function],
              "index": 2,
              "name": "name",
              "strings": Array [
                "",
                "",
              ],
              "type": 1,
            },
          ],
        },
        "_$AM": [Circular],
        "_$AN": undefined,
        "u": Array [
          R {
            "_$AH": "0",
            "_$AM": [Circular],
            "_$AN": undefined,
            "element": <button
              class="inline-block px-6 font-medium text-xs leading-tight uppercase focus:outline-none focus:ring-0 transition duration-150 ease-in-out
      py-2.5 bg-blue-600 text-white shadow-md hover:bg-blue-700 hover:shadow-lg focus:bg-blue-700 focus:shadow-lg  active:bg-blue-800 active:shadow-lg
      rounded
      undefined
      undefined
    "
              part="custom-button"
              role="button"
              tabindex="0"
              type="button"
            >
              
        
              <!--?lit$7157807873$-->
              
        
              <slot
                name=""
              />
              
      
            </button>,
            "name": "tabindex",
            "options": Object {
              "host": [Circular],
              "isConnected": true,
              "renderBefore": null,
            },
            "type": 1,
          },
          R {
            "_$AH": "button",
            "_$AM": [Circular],
            "_$AN": undefined,
            "element": <button
              class="inline-block px-6 font-medium text-xs leading-tight uppercase focus:outline-none focus:ring-0 transition duration-150 ease-in-out
      py-2.5 bg-blue-600 text-white shadow-md hover:bg-blue-700 hover:shadow-lg focus:bg-blue-700 focus:shadow-lg  active:bg-blue-800 active:shadow-lg
      rounded
      undefined
      undefined
    "
              part="custom-button"
              role="button"
              tabindex="0"
              type="button"
            >
              
        
              <!--?lit$7157807873$-->
              
        
              <slot
                name=""
              />
              
      
            </button>,
            "name": "role",
            "options": Object {
              "host": [Circular],
              "isConnected": true,
              "renderBefore": null,
            },
            "type": 1,
          },
          R {
            "_$AH": "inline-block px-6 font-medium text-xs leading-tight uppercase focus:outline-none focus:ring-0 transition duration-150 ease-in-out
      py-2.5 bg-blue-600 text-white shadow-md hover:bg-blue-700 hover:shadow-lg focus:bg-blue-700 focus:shadow-lg  active:bg-blue-800 active:shadow-lg
      rounded
      undefined
      undefined
    ",
            "_$AM": [Circular],
            "_$AN": undefined,
            "element": <button
              class="inline-block px-6 font-medium text-xs leading-tight uppercase focus:outline-none focus:ring-0 transition duration-150 ease-in-out
      py-2.5 bg-blue-600 text-white shadow-md hover:bg-blue-700 hover:shadow-lg focus:bg-blue-700 focus:shadow-lg  active:bg-blue-800 active:shadow-lg
      rounded
      undefined
      undefined
    "
              part="custom-button"
              role="button"
              tabindex="0"
              type="button"
            >
              
        
              <!--?lit$7157807873$-->
              
        
              <slot
                name=""
              />
              
      
            </button>,
            "name": "class",
            "options": Object {
              "host": [Circular],
              "isConnected": true,
              "renderBefore": null,
            },
            "type": 1,
          },
          O {
            "_$AA": <!--?lit$7157807873$-->,
            "_$AB": 
        ,
            "_$AH": Symbol(lit-nothing),
            "_$AM": [Circular],
            "_$AN": undefined,
            "_$Cm": true,
            "options": Object {
              "host": [Circular],
              "isConnected": true,
              "renderBefore": null,
            },
            "type": 2,
          },
          R {
            "_$AH": undefined,
            "_$AM": [Circular],
            "_$AN": undefined,
            "element": <slot
              name=""
            />,
            "name": "name",
            "options": Object {
              "host": [Circular],
              "isConnected": true,
              "renderBefore": null,
            },
            "type": 1,
          },
        ],
      },
      "_$AM": undefined,
      "_$AN": undefined,
      "_$Cm": true,
      "options": Object {
        "host": [Circular],
        "isConnected": true,
        "renderBefore": null,
      },
      "type": 2,
    },
  },
}
`;
